### PR TITLE
kernel, ksud: collect binaries into /data/adb/ksu

### PR DESCRIPTION
**Pull request by skbeh**

### CODE REVIEW

**Comment by tiann** [comment]
When the directory is created? `/data/adb` is created by Android, if you use its subdirectory, you should create it.

**Comment by skbeh** [comment]
> When the directory is created? `/data/adb` is created by Android, if you use its subdirectory, you should create it.

It is created before usage.

**Review by tiann — CHANGES_REQUESTED** [review]

File: kernel/allowlist.c [hunk 1]
```diff
@@ -21,7 +21,7 @@ struct perm_data {
 
 static struct list_head allow_list;
 
-#define KERNEL_SU_ALLOWLIST "/data/adb/.ksu_allowlist"
+#define KERNEL_SU_ALLOWLIST "/data/adb/ksu/allowlist"
```
[the remark this answers is not in the document]
Comment (skbeh) [reply kernel/allowlist.c:24]: I think that the one who wants to delete it will delete it anyway, and magisk doesn't use hidden file.

**Comment by tiann** [comment]
I will test it with my device later

**Comment by tiann** [comment]
Please keep `/data/adb/ksud` the same, this command is useful for developers, we don't like to type more words.
Other modification is OK.

**Comment by skbeh** [comment]
> Please keep `/data/adb/ksud` the same, this command is useful for developers, we don't like to type more words. Other modification is OK.

Would you like to move it to somewhere in `PATH`?

**Comment by skbeh** [comment]
Or we can modify `PATH` for the commands which are executed by `su`.

**Comment by tiann** [comment]
> > Please keep `/data/adb/ksud` the same, this command is useful for developers, we don't like to type more words. Other modification is OK.
> 
> Would you like to move it to somewhere in `PATH`?

We don't want to put it in $PATH, it is an internal implementation, not an API.

**Comment by skbeh** [comment]
Would you like to ship `busybox` to gain some compatibility with magisk?

**Comment by tiann** [comment]
> Would you like to embed `busybox` to gain some compatibility with magisk?

we should do that

**Comment by skbeh** [comment]
@tiann BusyBox is licensed under GPLv2, so embedding it breaks the license. Also the linux kernel is licensed under GPLv2, so linking kernelsu with it also breaks license.

**Comment by skbeh** [comment]
I suggest to license kernelsu under GPLv2 or GPLv2-or-later.

**Comment by tiann** [comment]
KernelSU is GPL-3

**Comment by skbeh** [comment]
> KernelSU is GPL-3

GPLv3 is not compatible with GPLv2.

**Comment by tiann** [comment]
Magisk is also GPL3

**Comment by skbeh** [comment]
> Magisk is also GPL3

Because magisk doesn't link with Linux kernel or GPLv2 libraries.

**Comment by tiann** [comment]
GPLv2-or-later should be Ok？Magisk is GPL3 and we use its reserprop

**Comment by skbeh** [comment]
> GPLv2-or-later should be Ok

I think it should be. However, I am not a lawyer, so I can not give legal advices.

**Comment by tiann** [comment]
I consulted the layer, kernel is gpl2 without later, so the kernel part of KernelSU must be gpl2 without later.
So, I will relicense it under gpl2.

**Comment by skbeh** [comment]
The binaries are directly taken from Magisk v25.2 apk.

**Review by tiann — CHANGES_REQUESTED** [review]

File: userspace/ksud/src/defs.rs [hunk 2]
```diff
@@ -1,7 +1,9 @@
 use const_format::concatcp;
 
-pub const DAEMON_PATH: &str = "/data/adb/ksud";
 pub const WORKING_DIR: &str = "/data/adb/ksu/";
+pub const BINARY_DIR: &str = concatcp!(WORKING_DIR, "bin/");
+
+pub const DAEMON_PATH: &str = concatcp!(WORKING_DIR, "ksud");
```
[the remark this answers is not in the document]
Comment (skbeh) [reply userspace/ksud/src/defs.rs:6]: Why?

[the remark this answers is not in the document]
Comment (tiann) [reply userspace/ksud/src/defs.rs:6]: it is `/data/adb/ksu/ksud` but `/data/adb/ksud` in kernel

File: userspace/ksud/src/event.rs [hunk 3]
```diff
@@ -179,6 +179,7 @@ pub fn install() -> Result<()> {
     let src = "/proc/self/exe";
     let dst = defs::DAEMON_PATH;
 
+    ensure_dir_exists(dst)?;
```
[the remark this answers is not in the document]
Comment (skbeh) [reply userspace/ksud/src/event.rs]: It should be checked in case `/data/adb` is deleted.

**Review by tiann — APPROVED** [review]

**Pull request merged**